tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc

```diff
@@ -1946,6 +1946,96 @@ static LogicalResult Verify(DataFormatVecPermuteOp op) {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// DivNoNanOp
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// Canonicalization template for tf.DivNoNan and tf.MulNoNan:
+/// If the op is tf.DivNoNan and the divisor is a constant tensor (with all the
+/// elements of any allowed type: float or complex), rewrite the op to the
+/// divisor if all the elements of the divisor are zero and to tf.Div if all the
+/// elements of the divisor are non-zero.
+
+/// Similarly, if the op is tf.MulNoNan and the multiplier is a constant tensor
+/// (with all the elements of any allowed type: float or complex), rewrite the
+/// op to the multiplier if all the elements of the multiplier are zero and to
+/// tf.Mul if all the elements of the multiplier are non-zero.
+
+/// Replace the given op with an op of type `RetT`. Upon calling
+/// DivNoNanOrMulNoNanConstantY for canonicalizing tf.DivNoNan, TF::DivOp is
+/// passed as the second argument and for canonicalizing tf.MulNoNan, TF::MulOp
+/// is passed as the second argument.
+template <typename OpT, typename RetT>
+class DivNoNanOrMulNoNanConstantY : public OpRewritePattern<OpT> {
+  using OpRewritePattern<OpT>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(OpT op,
+                                PatternRewriter& rewriter) const override {
+    static_assert(
+        llvm::is_one_of<OpT, DivNoNanOp, MulNoNanOp>::value,
+        "only canonicalization of tf.DivNoNan and tf.MulNoNan is supported");
+
+    // A lambda function named `numZerosInAttr` is used for returning the
+    // total number of zero elements (float/complex type) in `attr`.
+    auto numZerosInAttr = [](ElementsAttr attr, bool complexElements) {
+      unsigned numZeros = 0;
+      if (!complexElements) {
+        for (const auto val : attr.getValues<APFloat>()) {
+          if (val.isZero()) numZeros++;
+        }
+      } else {
+        for (const auto val : attr.getValues<std::complex<APFloat>>()) {
+          // Note that when `val` is of complex type, it is zero iff both
+          // its real and imaginary parts are zero.
+          if (val.real().isZero() && val.imag().isZero()) numZeros++;
+        }
+      }
+      return numZeros;
+    };
```

Comment:
Looking at the usage, you don't really need the number of zeros: you only need to know if we either:
- have a splat of a 0.0 (I don't think you need to iterate for this, but even if you do, you only need to stop on the first non 0).
- have no zero: you need to iterate but you don't need to count, you can stop on the first zero.

Reply:
> Looking at the usage, you don't really need the number of zeros: you only need to know if we either:
> 
> * have a splat of a 0.0 (I don't think you need to iterate for this, but even if you do, you only need to stop on the first non 0).
> * have no zero: you need to iterate but you don't need to count, you can stop on the first zero.

That's right! Thanks for this suggestion. I will modify the code accordingly.

Reply:
> Looking at the usage, you don't really need the number of zeros: you only need to know if we either:
> 
> * have a splat of a 0.0 (I don't think you need to iterate for this, but even if you do, you only need to stop on the first non 0).
> * have no zero: you need to iterate but you don't need to count, you can stop on the first zero.

@joker-eph , I have made the required changes! Let me know if they seem fine!